app/javascript/packs/followButtons.js

```diff
@@ -0,0 +1,366 @@
+        break;
+      case 'login':
+        addButtonFollowText(matchingButton, style);
+        break;
+      case 'self':
+        updateUserOwnFollowButton(matchingButton);
+        break;
+      default:
+        updateFollowingButton(matchingButton, style);
+    }
+  });
+}
+
+/**
+ * Set the Follow button's UI to the 'following' state
+ *
+ * @param {HTMLElement} button The Follow button to be updated
+ * @param {string} style Style of the follow button (e.g. 'small')
+ */
+function updateFollowingButton(button, style) {
+  button.dataset.verb = 'follow';
+  addButtonFollowingText(button, style);
+  button.classList.remove('crayons-btn--primary');
+  button.classList.remove('crayons-btn--secondary');
+  button.classList.add('crayons-btn--outlined');
+}
+
+/**
+ * Update the UI of the given button to the user's own button - i.e. 'Edit profile'
+ *
+ * @param {HTMLElement} button The Follow button to be updated
+ */
+function updateUserOwnFollowButton(button) {
+  button.dataset.verb = 'self';
+  button.textContent = 'Edit profile';
+}
+
+/**
+ * Update the UI of the given button to the 'follow' or 'follow-back' state
+ *
+ * @param {HTMLElement} button The Follow button to be updated
+ * @param {string} newState The new follow state of the button
+ * @param {Object} buttonInfo The parsed info object obtained from the button's dataset
+ * @param {string} buttonInfo.style The style of the follow button (e.g 'small')
+ * @param {string} buttonInfo.followStyle The crayons button variant (e.g 'primary')
+ */
+function updateFollowButton(button, newState, buttonInfo) {
+  const { style, followStyle } = buttonInfo;
+
+  button.dataset.verb = 'unfollow';
+  button.classList.remove('crayons-btn--outlined');
+
+  if (followStyle === 'primary') {
+    button.classList.add('crayons-btn--primary');
+  } else if (followStyle === 'secondary') {
+    button.classList.add('crayons-btn--secondary');
+  }
+
+  const nextButtonStyle = newState === 'follow-back' ? newState : style;
+  addButtonFollowText(button, nextButtonStyle);
+}
+
+/**
+ * Checks a click event's target, and if it is a follow button, triggers the appropriate follow action
+ *
+ * @param {HTMLElement} target The target of the click event
+ */
+function handleFollowButtonClick({ target }) {
+  if (
+    target.classList.contains('follow-action-button') ||
+    target.classList.contains('follow-user')
+  ) {
+    const userStatus = document.body.getAttribute('data-user-status');
+    if (userStatus === 'logged-out') {
+      showLoginModal();
+      return;
+    }
+
+    optimisticallyUpdateButtonUI(target);
+
+    const { verb } = target.dataset;
+
+    if (verb === 'self') {
+      window.location.href = '/settings';
+      return;
+    }
+
+    const buttonDataInfo = JSON.parse(target.dataset.info);
+    const formData = new FormData();
+    formData.append('followable_type', buttonDataInfo.className);
+    formData.append('followable_id', buttonDataInfo.id);
```

Comment:
Refactor (non-blocking): Prefer destructuring off of the object created from `JSON.parse`.

```suggestion
    const { className, id }  = JSON.parse(target.dataset.info);
    const formData = new FormData();
    formData.append('followable_type', className);
    formData.append('followable_id', id);
```